cirq/protocols/apply_unitary.py

```diff
@@ -86,28 +90,60 @@ def __init__(self,
                 dtype as the target tensor.
             axes: Which axes the unitary effect is being applied to (e.g. the
                 qubits that the gate is operating on).
+            qid_shape: A tuple of the dimensions of each qid in the state
+                `target_tensor`.
         """
+        self.axes = tuple(axes)
+        if qid_shape is None:
+            # Guess what the qid_shape would be for qubits.
+            min_num_qubits = max(self.axes, default=-1) + 1
+            qid_shape = (2,) * min_num_qubits  # Might be too short
         self.target_tensor = target_tensor
         self.available_buffer = available_buffer
-        self.axes = tuple(axes)
+        self.qid_shape = qid_shape
 
     @staticmethod
-    def default(num_qubits: int) -> 'ApplyUnitaryArgs':
-        """A default instance starting in state |0⟩."""
+    def default(num_qubits: Optional[int] = None,
+                *,
+                qid_shape: Optional[Tuple[int, ...]] = None
+               ) -> 'ApplyUnitaryArgs':
+        """A default instance starting in state |0⟩.
+
+        Specify exactly one argument.
+
+        Args:
+            num_qubits: The number of qubits to make space for in the state.
+            qid_shape: The shape of the state, specifying the dimension of each
+                qid."""
+        if (num_qubits is None) == (qid_shape is None):
+            raise TypeError(
+                'Specify either the num_qubits or qid_shape argument.')
+        if num_qubits is not None:
+            qid_shape = (2,) * num_qubits
+        assert qid_shape is not None, "Can't be None here"  # Satisfy mypy
+        num_qubits = len(qid_shape)
         state = linalg.one_hot(index=(0,) * num_qubits,
-                               shape=(2,) * num_qubits,
+                               shape=qid_shape,
                                dtype=np.complex128)
-        return ApplyUnitaryArgs(state, np.empty_like(state), range(num_qubits))
-
-    def subspace_index(self, little_endian_bits_int: int
-                       ) -> Tuple[Union[slice, int, 'ellipsis'], ...]:
+        return ApplyUnitaryArgs(state, np.empty_like(state), range(num_qubits),
+                                qid_shape)
+
+    def subspace_index(
```

Comment:
This method is specifically about "I want to give an integer index", so I would not add a tuple variant argument.

Instead of that, we should use the shape to pull apart the integer in a little endian fashion. For example `5 = 2*1 + 1*3` so 5 corresponds to [2, 1] when the shape is (*, 3).

We can separately consider adding an alternate method that scatters a tuple of indices over the axis indices.

Reply:
Removed tuple argument.